src/System.Management.Automation/engine/parser/TypeResolver.cs

```diff
@@ -302,10 +321,28 @@ internal static Type ResolveTypeNameWithContext(TypeName typeName, out Exception
                 return typeName._typeDefinitionAst.Type;
             }
 
-            result = ResolveTypeNameWorker(typeName, currentScope, typeResolutionState.assemblies, typeResolutionState, /* reportAmbiguousException */ true, out exception);
+            var assemList = assemblies ?? ClrFacade.GetAssemblies(typeResolutionState, typeName);
+            // A set of assembly names that we have searched from but found no matching type.
+            // This set will be updated within ResolveTypeNameWorker.
+            var searchedAssemblies = new HashSet<string>();
```

Comment:
Good point. The `searchedAssemblies` set exists only for optimization purpose. I can skip adding dynamic assembly names to this set, so we will search repeatedly for dynamic assemblies, but assemblies held by `context.AssemblyCache` should rarely have dynamic assemblies.